contracts/farm/LiquidityMiningV1.sol

```diff
@@ -0,0 +1,708 @@
+		bool _updateAllFlag
+	) external onlyAuthorized {
+		require(_poolTokens.length == _allocationPoints.length, "Arrays mismatch");
+
+		if (_updateAllFlag) {
+			updateAllPools();
+		}
+		uint256 length = _poolTokens.length;
+		for (uint256 i = 0; i < length; i++) {
+			if (!_updateAllFlag) {
+				updatePool(_poolTokens[i]);
+			}
+			_updateToken(_poolTokens[i], _allocationPoints[i]);
+		}
+	}
+
+	/**
+	 * @notice returns reward multiplier over the given _from to _to block
+	 * @param _from the first block for a calculation
+	 * @param _to the last block for a calculation
+	 */
+	function _getPassedBlocksWithBonusMultiplier(uint256 _from, uint256 _to) internal view returns (uint256) {
+		if (_from < startBlock) {
+			_from = startBlock;
+		}
+		if (endBlock > 0 && _to > endBlock) {
+			_to = endBlock;
+		}
+		if (_to <= bonusEndBlock) {
+			return _to.sub(_from).mul(BONUS_BLOCK_MULTIPLIER);
+		} else if (_from >= bonusEndBlock) {
+			return _to.sub(_from);
+		} else {
+			return bonusEndBlock.sub(_from).mul(BONUS_BLOCK_MULTIPLIER).add(_to.sub(bonusEndBlock));
+		}
+	}
+
+	function _getUserAccumulatedReward(uint256 _poolId, address _user) internal view returns (uint256) {
+		PoolInfo storage pool = poolInfoList[_poolId];
+		UserInfo storage user = userInfoMap[_poolId][_user];
+
+		uint256 accumulatedRewardPerShare = pool.accumulatedRewardPerShare;
+		uint256 poolTokenBalance = pool.poolToken.balanceOf(address(this));
+		if (block.number > pool.lastRewardBlock && poolTokenBalance != 0) {
+			(, uint256 accumulatedRewardPerShare_) = _getPoolAccumulatedReward(pool);
+			accumulatedRewardPerShare = accumulatedRewardPerShare.add(accumulatedRewardPerShare_);
+		}
+		return user.amount.mul(accumulatedRewardPerShare).div(PRECISION).sub(user.rewardDebt);
+	}
+
+	/**
+	 * @notice returns accumulated reward
+	 * @param _poolToken the address of pool token
+	 * @param _user the user address
+	 */
+	function getUserAccumulatedReward(address _poolToken, address _user) external view returns (uint256) {
+		uint256 poolId = _getPoolId(_poolToken);
+		return _getUserAccumulatedReward(poolId, _user);
+	}
+
+	/**
+	 * @notice returns estimated reward
+	 * @param _poolToken the address of pool token
+	 * @param _amount the amount of tokens to be deposited
+	 * @param _duration the duration of liquidity providing in seconds
+	 */
+	function getEstimatedReward(
+		address _poolToken,
+		uint256 _amount,
+		uint256 _duration
+	) external view returns (uint256) {
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		uint256 start = block.number;
+		uint256 end = start.add(_duration.div(SECONDS_PER_BLOCK));
+		(, uint256 accumulatedRewardPerShare) = _getPoolAccumulatedReward(pool, _amount, start, end);
+		return _amount.mul(accumulatedRewardPerShare).div(PRECISION);
+	}
+
+	/**
+	 * @notice Updates reward variables for all pools.
+	 * @dev Be careful of gas spending!
+	 */
+	function updateAllPools() public {
```

Comment:
this method should be "onlyBeforeMigrationGracePeriodFinished" because it can be called by any user